test/policy.bats

```diff
@@ -97,3 +109,98 @@ SANDBOX_CONFIG="$TESTDATA/sandbox_config.json"
+	jq '.image.image = "'"$IMAGE_DIGEST"'" | .image.user_specified_image = "'"$UNSIGNED_IMAGE"'"' "$TESTDATA/container_config.json" > "$CTR_CONFIG"
+
+	run ! crictl create "$POD_ID" "$CTR_CONFIG" "$TESTDATA/sandbox_config.json"
+
+	[[ "$output" == *"SignatureValidationFailed"* ]]
+}
+
+@test "allow signed image with restrictive policy on container creation3 if already pulled (by ID)" {
+	start_crio
+	crictl pull "$SIGNED_IMAGE"
+	IMAGE_ID=$(crictl images -q "$SIGNED_IMAGE")
+	stop_crio_no_clean
+
+	SIGNATURE_POLICY="$RESTRICTIVE_POLICY" start_crio
+	POD_ID=$(crictl runp "$TESTDATA/sandbox_config.json")
+	CTR_CONFIG="$TESTDIR/config.json"
+	jq '.image.image = "'"$IMAGE_ID"'" | .image.user_specified_image = "'"$SIGNED_IMAGE"'"' "$TESTDATA/container_config.json" > "$CTR_CONFIG"
+
+	# Testing for container start failed not because of the signature, but of
+	# the missing command executable
+	run ! crictl create "$POD_ID" "$CTR_CONFIG" "$TESTDATA/sandbox_config.json"
+	[[ "$output" == *"unable to start container process"* ]]
+}
+
+@test "deny unsigned image with restrictive policy on container creation4 if already pulled (by ID)" {
+	start_crio
+	crictl pull "$UNSIGNED_IMAGE"
+	IMAGE_ID=$(crictl images -q "$UNSIGNED_IMAGE")
+	stop_crio_no_clean
+
+	SIGNATURE_POLICY="$RESTRICTIVE_POLICY" start_crio
+	POD_ID=$(crictl runp "$TESTDATA/sandbox_config.json")
+	CTR_CONFIG="$TESTDIR/config.json"
+	jq '.image.image = "'"$IMAGE_ID"'" | .image.user_specified_image = "'"$UNSIGNED_IMAGE"'"' "$TESTDATA/container_config.json" > "$CTR_CONFIG"
+
+	run ! crictl create "$POD_ID" "$CTR_CONFIG" "$TESTDATA/sandbox_config.json"
+
+	[[ "$output" == *"SignatureValidationFailed"* ]]
+}
+
+@test "allow signed image with restrictive policy on container creation5 if already pulled (by tag)" {
+	start_crio
+	crictl pull "$SIGNED_IMAGE"
+	stop_crio_no_clean
+
+	SIGNATURE_POLICY="$RESTRICTIVE_POLICY" start_crio
+	POD_ID=$(crictl runp "$TESTDATA/sandbox_config.json")
+	CTR_CONFIG="$TESTDIR/config.json"
+	jq '.image.image = "'"$SIGNED_IMAGE"'" | .image.user_specified_image = "'"$SIGNED_IMAGE"'"' "$TESTDATA/container_config.json" > "$CTR_CONFIG"
+
+	# Testing for container start failed not because of the signature, but of
+	# the missing command executable
+	run ! crictl create "$POD_ID" "$CTR_CONFIG" "$TESTDATA/sandbox_config.json"
+	[[ "$output" == *"unable to start container process"* ]]
+}
+
+@test "deny unsigned image with restrictive policy on container creation6 if already pulled (by tag)" {
+	start_crio
+	crictl pull "$UNSIGNED_IMAGE"
+	stop_crio_no_clean
+
+	SIGNATURE_POLICY="$RESTRICTIVE_POLICY" start_crio
+	POD_ID=$(crictl runp "$TESTDATA/sandbox_config.json")
+	CTR_CONFIG="$TESTDIR/config.json"
+	jq '.image.image = "'"$UNSIGNED_IMAGE"'" | .image.user_specified_image = "'"$UNSIGNED_IMAGE"'"' "$TESTDATA/container_config.json" > "$CTR_CONFIG"
+
+	run ! crictl create "$POD_ID" "$CTR_CONFIG" "$TESTDATA/sandbox_config.json"
+
+	[[ "$output" == *"SignatureValidationFailed"* ]]
+}
```

Comment:
Other interesting tests:
- Preparation:
  - Tag the same image as nameSigned:tagSigned and nameUnsigned:tagUnsigned. Sign the …Signed name. 
  - Using a permissive policy, pull both nameSigned:tagSigned, and nameUnsigned:tagUnsigned. Sanity-check that the local image ID is the same.
- Then, using a signature-enforcing policy:
  - (Start a container using UserSpecifiedImage: nameSigned:tagSigned, both by tag and by ID: should work)
  - Start a container using UserSpecifiedImage: nameUnsigned:tagUnsigned, by ID (mirroring a Kubelet that does ImageStatus(…Unsigned) and does not trigger a pull): should fail
  - Remove the nameUnsigned:tagUnsigned tag; start a container using UserSpecifiedImage: nameUnsigned:tagUnsigned, by ID (mirroring a Kubelet that does imageStatus, racing with a concurrent pull of a new version of …Unsigned moving the tag): should fail
  - Extra credit, not essential? Remove the  Remove the nameSigned:tagSigned tag; start a container using UserSpecifiedImage: nameSigned:tagSigned, by ID: should succeed.

I think the last one would currently fail because the code looks up in local storage by repo. OTOH this interacts with the “use incorrect digest” concern — we might have settled on this lookup approach in previous discussions.

Some of the “should fail” would have incorrectly succeeded in yesterday’s version of this PR.

Reply:
With the latest changes all those existing tests pass, 
```
1..15
ok 1 accept unsigned image with default policy
ok 2 deny unsigned image with restrictive policy
ok 3 deny unsigned image with restrictive policy if already pulled
ok 4 accept signed image with default policy
ok 5 accept signed image with restrictive policy
ok 6 accept unsigned image with not existing namespace policy
ok 7 accept unsigned image with higher priority namespace policy
ok 8 deny unsigned image with higher priority namespace policy
ok 9 accept signed image with higher priority namespace policy
ok 10 allow signed image with restrictive policy on container creation1 (fresh pull)
ok 11 deny unsigned image with restrictive policy on container creation2 (fresh pull)
ok 12 allow signed image with restrictive policy on container creation3 if already pulled (by ID)
ok 13 deny unsigned image with restrictive policy on container creation4 if already pulled (by ID)
ok 14 allow signed image with restrictive policy on container creation5 if already pulled (by tag)
ok 15 deny unsigned image with restrictive policy on container creation6 if already pulled (by tag)
``` 

but I am not sure what is going on with the mocked tests, https://github.com/cri-o/cri-o/pull/8212#issuecomment-2322025295